std/machines/hash/keccak.asm

```diff
@@ -0,0 +1,544 @@
+
+    //     // Note: D is inlined, not stored in the witness.
+    //     /// ```ignore
+    //     /// A'[x, y] = xor(A[x, y], D[x])
+    //     ///          = xor(A[x, y], C[x - 1], ROT(C[x + 1], 1))
+    //     /// ```
+    //     pub a_prime: [[[T; 64]; 5]; 5],
+
+    //     /// ```ignore
+    //     /// A''[x, y] = xor(B[x, y], andn(B[x + 1, y], B[x + 2, y])).
+    //     /// ```
+    //     pub a_prime_prime: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     /// The bits of `A''[0, 0]`.
+    //     pub a_prime_prime_0_0_bits: [T; 64],
+
+    //     /// ```ignore
+    //     /// A'''[0, 0, z] = A''[0, 0, z] ^ RC[k, z]
+    //     /// ```
+    //     pub a_prime_prime_prime_0_0_limbs: [T; U64_LIMBS],
+    // }
+
+    pol commit export;
+    pol commit preimage[100]; // 5 * 5 * 4
+    pol commit a[100]; // 5 * 5 * 4
+    pol commit c[320]; // 5 * 64
+    pol commit c_prime[320]; // 5 * 64
+    pol commit a_prime[1600]; // 5 * 5 * 64
+    pol commit a_prime_prime[100]; // 5 * 5 * 4 
+    pol commit a_prime_prime_0_0_bits[64]; // 64
+    pol commit a_prime_prime_prime_0_0_limbs[4]; // 4
+
+    // eval_round_flags(builder);
+
+    // Initially, the first step flag should be 1 while the others should be 0.
+    // builder.when_first_row().assert_one(local.step_flags[0]);
+    // for i in 1..NUM_ROUNDS {
+    //     builder.when_first_row().assert_zero(local.step_flags[i]);
+    // }
+    // for i in 0..NUM_ROUNDS {
+    //     let current_round_flag = local.step_flags[i];
+    //     let next_round_flag = next.step_flags[(i + 1) % NUM_ROUNDS];
+    //     builder
+    //         .when_transition()
+    //         .assert_eq(next_round_flag, current_round_flag);
+    // }
+
+    let step_flags: col[NUM_ROUNDS] = array::new(NUM_ROUNDS, |i| |row| if row % NUM_ROUNDS == i { 1 } else { 0 } );
+
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    pol constant first_step(i) { if i % NUM_ROUNDS == 0 { 1 } else { 0 } }; // same as step_flags[0]
+    pol constant final_step(i) { if i % NUM_ROUNDS == NUM_ROUNDS - 1 { 1 } else { 0 } }; // same as step_flags[NUM_ROUNDS - 1]
+    pol constant not_final_step(i) { if i % NUM_ROUNDS == NUM_ROUNDS - 1 { 0 } else { 1 } };
+
+    // // If this is the first step, the input A must match the preimage.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(first_step)
+    //                 .assert_eq(local.preimage[y][x][limb], local.a[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::new(100, |i| first_step * (preimage[i] - a[i]) = 0);
```

Comment:
```suggestion
    array::zip(preimage, a, |p, a_el| first_step * (p - a_el) = 0);
```

(the naming of a_el is unfortunate, but maybe the name for `a` could also be longer?

Reply:
Agree on simplification. Keeping `a` naming for now to be consistent with Plonky3, but will change in the end.